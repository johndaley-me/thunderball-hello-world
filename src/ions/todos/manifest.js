module.exports = {
  name: 'todos',
  browser: {
    page: {
      createRoutes: './createRoutes',
      path: '/todos*',
      injectors: [
        '../../appInjectors',
        './injectors',
      ],
    },
  },
};
