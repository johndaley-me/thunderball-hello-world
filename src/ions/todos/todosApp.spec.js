/**
 * Integration tests in here include setup of routes and store similar to what
 * thunderball-client pageBuilder does
 */
/* eslint-disable global-require, import/no-dynamic-require, import/no-extraneous-dependencies */
import React from 'react';
import { browserHistory, match } from 'react-router';
import { mount } from 'enzyme';
import Shell from 'thunderball-client/lib/render/Shell';
import getStoreAndRoutes from 'thunderball-client/lib/render/getStoreAndRoutes';
import manifest from './manifest';

jest.mock('react-helmet');

// parse manifest
const injectors = manifest.browser.page.injectors.map(path => require(path));
const createRoutes = require(manifest.browser.page.createRoutes);
const pageProps = manifest.browser.page.props || {};

const defaultLocale = 'en';

const initialState = {
  todos: [
    {
      text: 'Test AddTodo',
      completed: false,
      id: 0,
    },
    {
      text: 'Test AddTodo 2',
      completed: true,
      id: 1,
    },
  ],
  visibilityFilter: 'SHOW_ALL',
};

const setup = () => {
  const { store, routes, history } = getStoreAndRoutes(
    initialState,
    createRoutes,
    browserHistory,
    { adjustUrlOnReplay: false },
    injectors,
    pageProps);
  history.push('/todos');
  const props = { injectors, store, history, pageProps, routes, defaultLocale };
  const wrapper = mount(<Shell {...props} />);

  return {
    props,
    store,
    wrapper,
  };
};

describe('todos app integration', () => {
  describe('server side rendering', () => {
    it('renders /todos', (done) => {
      expect.assertions(2);
      const { store, routes, history } = getStoreAndRoutes(
        initialState,
        createRoutes,
        browserHistory,
        { adjustUrlOnReplay: false },
        injectors,
        pageProps);
      expect(store.getState()).toMatchSnapshot();
      const shellProps = { injectors, store, history, pageProps, routes, defaultLocale };
      match({ history, routes, location: '/todos' }, (error, redirectLocation, renderProps) => {
        const wrapper = mount(<Shell {...shellProps} renderProps={renderProps} />);
        expect(wrapper.render()).toMatchSnapshot();
        done();
      });
    });
  });
  describe('client side', () => {
    beforeEach(() => {
      process.env.IS_BROWSER = 'true';
    });
    afterEach(() => {
      delete process.env.IS_BROWSER;
    });
    it('renders /todos', () => {
      const { wrapper } = setup();
      expect(wrapper.render()).toMatchSnapshot();
    });
    it('adds todo', () => {
      const { wrapper } = setup();
      expect(wrapper.find('Todo')).toHaveLength(2);
      const form = wrapper.find('form');
      const input = form.find('input');
      const button = form.find('button');
      const value = 'test - go to the store';
      input.simulate('change', { target: { value } });
      expect(input.prop('value')).toEqual(value);
      button.simulate('submit');
      expect(wrapper.find('Todo')).toHaveLength(3);
    });
    it('filters active todos', () => {
      const { wrapper } = setup();
      expect(wrapper.find('Todo')).toHaveLength(2);
      const button = wrapper.find('button').findWhere(x => x.text() === 'Active');
      expect(button).toHaveLength(1);
      button.simulate('click');
      expect(wrapper.find('Todo')).toHaveLength(1);
    });
  });
});
