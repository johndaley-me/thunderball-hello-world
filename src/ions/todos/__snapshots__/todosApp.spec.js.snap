// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos app integration client side renders /todos 1`] = `
<div>
  <div>
    Helmet Mock
    Object {
  "htmlAttributes": Object {
    "lang": "en",
  },
  "link": Array [
    Object {
      "href": Object {},
      "rel": "shortcut icon",
      "type": "image/png",
    },
    Object {
      "href": "//fonts.googleapis.com/css?family=Roboto:300,300italic,700,700italic",
      "rel": "stylesheet",
    },
    Object {
      "href": "//cdn.rawgit.com/necolas/normalize.css/master/normalize.css",
      "rel": "stylesheet",
    },
    Object {
      "href": "//cdn.rawgit.com/milligram/milligram/master/dist/milligram.min.css",
      "rel": "stylesheet",
    },
  ],
  "meta": Array [],
  "titleTemplate": "%s - Thunderball Hello World",
}
  </div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="column"
      >
        <div>
          <div>
            Helmet Mock
            Object {
  "title": "TODO",
}
          </div>
          <div>
            <h1>
              <span>
                TODO
              </span>
            </h1>
            <div>
              <form>
                <input
                  value=""
                />
                <button
                  type="submit"
                >
                  Add Todo
                </button>
              </form>
            </div>
            <div>
              <div>
                <input
                  id="todo-0"
                  type="checkbox"
                />
                <label
                  class="label-inline"
                  for="todo-0"
                >
                  Test AddTodo
                </label>
              </div>
              <div>
                <input
                  id="todo-1"
                  type="checkbox"
                />
                <label
                  class="label-inline"
                  for="todo-1"
                >
                  Test AddTodo 2
                </label>
              </div>
            </div>
            <div>
              <span>
                Show: 
              </span>
              <button
                disabled=""
                style="margin-left: 4px;"
              >
                All
              </button>
              <button
                style="margin-left: 4px;"
              >
                Active
              </button>
              <button
                style="margin-left: 4px;"
              >
                Completed
              </button>
            </div>
            <hr />
            <a
              href="/"
            >
              Home
            </a>
             
            &gt; Todos
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`todos app integration server side rendering renders /todos 1`] = `
Object {
  "config": Object {},
  "device": Object {
    "host": "",
    "platform": "",
  },
  "routing": Object {
    "locationBeforeTransitions": Object {
      "action": "POP",
      "hash": "",
      "key": null,
      "pathname": "/",
      "query": Object {},
      "search": "",
      "state": undefined,
    },
  },
  "todos": Array [
    Object {
      "completed": false,
      "id": 0,
      "text": "Test AddTodo",
    },
    Object {
      "completed": true,
      "id": 1,
      "text": "Test AddTodo 2",
    },
  ],
  "visibilityFilter": "SHOW_ALL",
}
`;

exports[`todos app integration server side rendering renders /todos 2`] = `
<div>
  <div>
    Helmet Mock
    Object {
  "htmlAttributes": Object {
    "lang": "en",
  },
  "link": Array [
    Object {
      "href": Object {},
      "rel": "shortcut icon",
      "type": "image/png",
    },
    Object {
      "href": "//fonts.googleapis.com/css?family=Roboto:300,300italic,700,700italic",
      "rel": "stylesheet",
    },
    Object {
      "href": "//cdn.rawgit.com/necolas/normalize.css/master/normalize.css",
      "rel": "stylesheet",
    },
    Object {
      "href": "//cdn.rawgit.com/milligram/milligram/master/dist/milligram.min.css",
      "rel": "stylesheet",
    },
  ],
  "meta": Array [],
  "titleTemplate": "%s - Thunderball Hello World",
}
  </div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="column"
      >
        <div>
          <div>
            Helmet Mock
            Object {
  "title": "TODO",
}
          </div>
          <div>
            <h1>
              <span>
                TODO
              </span>
            </h1>
            <div>
              <form>
                <input
                  value=""
                />
                <button
                  type="submit"
                >
                  Add Todo
                </button>
              </form>
            </div>
            <div>
              <div>
                <input
                  id="todo-0"
                  type="checkbox"
                />
                <label
                  class="label-inline"
                  for="todo-0"
                >
                  Test AddTodo
                </label>
              </div>
              <div>
                <input
                  id="todo-1"
                  type="checkbox"
                />
                <label
                  class="label-inline"
                  for="todo-1"
                >
                  Test AddTodo 2
                </label>
              </div>
            </div>
            <div>
              <span>
                Show: 
              </span>
              <button
                disabled=""
                style="margin-left: 4px;"
              >
                All
              </button>
              <button
                style="margin-left: 4px;"
              >
                Active
              </button>
              <button
                style="margin-left: 4px;"
              >
                Completed
              </button>
            </div>
            <hr />
            <a
              href="/"
            >
              Home
            </a>
             
            &gt; Todos
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
